### PR TITLE
Make provider uri getters return Urls

### PR DESCRIPTION
**Pull request by Korvox**
Having used inth_oauth2 in my OpenID implementation a lot this last month, I found the way auth_uri returns a Result to be really ergonomically cumbersome when the only reason it does so is because its parsing the `str` return from `auth_uri` on the Provider.

The tradeoff is that you need to either allocate Urls in the struct (as I did when updating the client tests) or you need something like lazy_static to construct the urls in the same style as the current string literals which is what I used in providers.

This change eliminates the need to have a Result on auth_uri, which means consumers who redirect to that uri can call it in their handlers without needing to worry about errors versus having to call it once, store the value, and get into some nasty business with keeping the state / scope / etc parameters consistent (or have to have error handling in their routes for this).

Long term this would probably be best served by `const_fn` to evaluate the Urls at compile time when it enters stable.

### CODE REVIEW

File: src/client/mod.rs [hunk 1]
```diff
@@ -128,7 +128,7 @@ impl<P: Provider> Client<P> {
         ]);
         let body = body.finish();
 
-        let mut response = http_client.post(self.provider.token_uri())?
+        let mut response = http_client.post(self.provider.token_uri().as_str())?
```
Comment (programble) [inline src/client/mod.rs]: Does this mean `http_client` re-parses the URL?

File: src/provider.rs [hunk 2]
```diff
@@ -99,6 +114,30 @@ pub struct Imgur;
...
     type Token = Bearer<Refresh>;
-    fn auth_uri(&self) -> &str { "https://api.imgur.com/oauth2/authorize" }
-    fn token_uri(&self) -> &str { "https://api.imgur.com/oauth2/token" }
+    fn auth_uri(&self) -> &Url { &IMGUR_AUTH_URI }
+    fn token_uri(&self) -> &Url { &IMGUR_TOKEN_URI }
 }
+
+#[test]
+fn google_urls() {
+    let prov = google::Web {};
```
Comment (programble) [inline src/provider.rs]: Braces like this are valid?

Comment (Korvox) [reply src/provider.rs]: Its empty struct initialization. Its the same syntax as initializing a struct with all pub fields like `let foo = Bar { baz: "hello" }` just without any member data.

Comment (programble) [reply src/provider.rs]: But why? The structs are unit structs.

File: src/client/mod.rs (hunk 1 above)
Comment (Korvox) [reply src/client/mod.rs]: Good catch, you can avoid the reparse by cloning Url - reqwest impls `IntoUrl` for its methods only for owned Urls. I committed that change.

File: src/provider.rs (hunk 2 above)
Comment (Korvox) [reply src/provider.rs]: I was probably thinking something like "remind anyone reading the test you can have fields in your Providers" but they are redundant so I got rid of them!

**Pull request merged**